src/cuckoocache.h

```diff
@@ -0,0 +1,451 @@
+            epoch_heuristic_counter = std::max(1u, std::max(epoch_size / 16,
+                        epoch_size - std::min(epoch_size, epoch_unused_count)));
+    }
+
+public:
+    /** You must always construct a cache with some elements via a subsequent
+     * call to setup or setup_bytes, otherwise operations may segfault.
+     */
+    cache() : table(), size(), collection_flags(0), epoch_flags(),
+    epoch_heuristic_counter(), epoch_size(), depth_limit(0), hash_function()
+    {
+    }
+
+    /** setup initializes the container to store no more than new_size
+     * elements. setup rounds down to a power of two size.
+     *
+     * setup should only be called once.
+     *
+     * @param new_size the desired number of elements to store
+     * @returns the maximum number of elements storable
+     **/
+    uint32_t setup(uint32_t new_size)
+    {
+        // depth_limit must be at least one otherwise errors can occur.
+        depth_limit = static_cast<uint8_t>(std::log2(static_cast<float>(std::max((uint32_t)2, new_size))));
+        size = 1 << depth_limit;
+        hash_mask = size-1;
+        table.resize(size);
+        collection_flags.setup(size);
+        epoch_flags.resize(size);
+        // Set to 45% as described above
+        epoch_size = std::max((uint32_t)1, (45 * size) / 100);
+        // Initially set to wait for a whole epoch
+        epoch_heuristic_counter = epoch_size;
+        return size;
+    }
+
+    /** setup_bytes is a convenience function which accounts for internal memory
+     * usage when deciding how many elements to store. It isn't perfect because
+     * it doesn't account for any overhead (struct size, MallocUsage, collection
+     * and epoch flags). This was done to simplify selecting a power of two
+     * size. In the expected use case, an extra two bits per entry should be
+     * negligible compared to the size of the elements.
+     *
+     * @param bytes the approximate number of bytes to use for this data
+     * structure.
+     * @returns the maximum number of elements storable (see setup()
+     * documentation for more detail)
+     */
+    uint32_t setup_bytes(size_t bytes)
+    {
+        return setup(bytes/sizeof(Element));
+    }
+
+    /** insert loops at most depth_limit times trying to insert a hash
+     * at various locations in the table via a variant of the Cuckoo Algorithm
+     * with two hash locations.
+     *
+     * It drops the last tried element if it runs out of depth before
+     * encountering an open slot.
+     *
+     * Thus
+     *
+     * insert(x);
+     * return contains(x, false);
+     *
+     * is not guaranteed to return true.
+     *
+     * @param e the element to insert
+     * @post one of the following: All previously inserted elements and e are
+     * now in the table, one previously inserted element is evicted from the
+     * table, the entry attempted to be inserted is evicted.
+     *
+     */
+    inline void insert(Element e)
```

Comment:
std::swap is just as slow for uint256, though. Since there is no dynamically-allocated memory we cant speed it up, really.